packages/@lwc/template-compiler/src/parser/index.ts

```diff
@@ -1109,15 +1219,33 @@ function parseSlot(
     ctx.addSeenSlot(name);
 
     if (alreadySeen) {
-        ctx.warnAtLocation(ParserDiagnostics.NO_DUPLICATE_SLOTS, location, [
-            name === '' ? 'default' : `name="${name}"`,
-        ]);
-    } else if (isInIteration(ctx)) {
+        // If slot name has been shared with a prior scoped slot, throw an error.
+        // Scoped slots do not allow duplicate or mixed slots
+        // https://rfcs.lwc.dev/rfcs/lwc/0118-scoped-slots-light-dom#restricting-ambigious-bindings
+        // https://rfcs.lwc.dev/rfcs/lwc/0118-scoped-slots-light-dom#invalid-usages
+        if (ctx.seenScopedSlots.has(name)) {
+            // Differentiate between mixed type or duplicate scoped slot
+            const errorInfo = isScopedSlot
+                ? ParserDiagnostics.NO_DUPLICATE_SCOPED_SLOT
+                : ParserDiagnostics.NO_MIXED_SLOT_TYPES;
+            ctx.throwAtLocation(errorInfo, location, [name === '' ? 'default' : `name="${name}"`]);
+        } else {
+            // for standard slots, preserve old behavior of warnings
+            ctx.warnAtLocation(ParserDiagnostics.NO_DUPLICATE_SLOTS, location, [
+                name === '' ? 'default' : `name="${name}"`,
+            ]);
+        }
+    } else if (!isScopedSlot && isInIteration(ctx)) {
+        // Scoped slots are allowed to be placed in iteration blocks
         ctx.warnAtLocation(ParserDiagnostics.NO_SLOTS_IN_ITERATOR, location, [
             name === '' ? 'default' : `name="${name}"`,
         ]);
     }
 
+    if (isScopedSlot) {
+        ctx.seenScopedSlots.add(name);
```

Comment:
Note that since the `alreadySeen` logic above reuses the standard slot name detection, the code in lines 1218 - 1219 will return false for `alreadySeen` when parsing the second `slotname1` slot in the above example and no error will be thrown. We'll just be adding the `name` to the `seenScopedSlots` set twice.

With the way `seenScopedSlots` is currently used, the code should work fine, but the parser is in an interesting brittle state when we parse the second slot in the above example: `alreadySeen === false`, because the other slot is in a separate conditional branch and isn't relevant to the current context, while `ctx.seenScopedSlots.has(name) === true`, because we don't make this distinction when managing the `seenScopedSlots` Set.

Reply:
@jye-sf 
I thought about this and added some test cases to come with this opinion. 
1. Basically, `this.seenScopedSlots` does not need to be if-block aware. The detection logic is handled by `ctx.hasSeenSlot`. The look up on `ctx.seenScopedSlots` is to just augment that information to produce an appropriate error/warning message. 

2. However, it can be made context aware by replicating the same was slot names are tracked. The scoped slots set will be a subset of the slot set and it does not bring much more else. I didn't see much value in doing that.

Let me know if 1 is okay or if you strongly prefer 2.

Reply:
I am okay with 1. It makes sense to not want to add a bunch of unnecessary duplicate code.

Maybe we can just rename `alreadySeen` to `seenInContext` or something like that; that might partially address the potential brittleness.